refactor: clean up -Wlogical-op warning in wallet/scriptpubkeyman.cpp

maskoficarus opened the pull request:
This is a quick patch that fixes #19912 . This change prevents a -Wlogical-op warning that occurs because we're treating a const int value as a boolean. There's no sense checking if a non-zero constant has a value, so I've removed the check.

#18836 also addresses the same warning, but has a larger scope and will require more review. This pull request will act as a patch to prevent this compile warning until 18836 is merged.

hebasto commented:
Thanks @maskoficarus !

https://github.com/bitcoin/bitcoin/issues/19912#issuecomment-688625883

kristapsk commented:
I'm not sure I get this. `VERSION_HD_CHAIN_SPLIT` is a constant, so what's the point of checking it's value here at all?

Commit:
refactor: Clean up -Wlogical-op warning

This commit fixes #19912 by removing a check that always returned true. That check was causing a -Wlogical-op warning because it treated a constant int as though it were a boolean.

maskoficarus commented:
> I'm not sure I get this. `VERSION_HD_CHAIN_SPLIT` is a constant, so what's the point of checking it's value here at all?

Good point. It shouldn't be needed.

MarcoFalke commented:
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

fanquake commented:
Thanks for squashing, however if we're going to merge this you'll need to use a commit message that is more useful/explanatory than the GitHub default. See [CONTRIBUTING.md#committing-patches](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches).

practicalswift commented:
@achow101 As the author of 415afcccd3e5583defdb76e3a280f48e98983301, would you mind reviewing this to make sure the the logic @maskoficarus suggests corresponds to what was originally intended? :)

MarcoFalke commented:
review ACK 95fedd33a23d6cb7542378afef229965f1ebfd68

Though, would be good to have instructions on how to reproduce the warning

kristapsk commented:
> would be good to have instructions on how to reproduce the warning

I was able to trigger this warning just by building with GCC 9.2.0 (on Gentoo Linux).

practicalswift commented:
@MarcoFalke gcc 9.3 which is the default compiler in Ubuntu 20.04 prints this warning :)

hebasto reviewed (approved):
ACK 95fedd33a23d6cb7542378afef229965f1ebfd68, tested on Linux Mint 20 (x86_64):

- master:
```
$ gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ git rev-parse HEAD
b99a1633b270e0e89479b2bb2ae19a8a8dc0fa05
$ make > /dev/null 
wallet/scriptpubkeyman.cpp: In member function ‘virtual bool LegacyScriptPubKeyMan::Upgrade(int, bilingual_str&)’:
wallet/scriptpubkeyman.cpp:455:55: warning: logical ‘and’ applied to non-boolean constant [-Wlogical-op]
  455 |     if (m_storage.CanSupportFeature(FEATURE_HD_SPLIT) && CHDChain::VERSION_HD_CHAIN_SPLIT) {
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

- PR:
````
$ git rev-parse HEAD
95fedd33a23d6cb7542378afef229965f1ebfd68
$ make > /dev/null 
# no output :)
